src/rgw/rgw_crypt.cc

```diff
@@ -0,0 +1,1314 @@
+      part_ofs += aligned_size;
+      off_t send_size = aligned_size - enc_begin_skip;
+      if (ofs + enc_begin_skip + send_size > end + 1) {
+        send_size = end + 1 - ofs - enc_begin_skip;
+      }
+      res = next->handle_data(data, enc_begin_skip, send_size);
+      enc_begin_skip = 0;
+      ofs += aligned_size;
+      if (res != 0)
+        return res;
+    }
+  }
+  return 0;
+}
+
+/**
+ * flush remainder of data to output
+ */
+int RGWGetObj_BlockDecrypt::flush() {
+  int res = 0;
+  size_t part_ofs = ofs;
+  size_t i = 0;
+  while (i<parts_len.size() && (part_ofs > parts_len[i])) {
+    part_ofs -= parts_len[i];
+    i++;
+  }
+  if (cache.length() > 0) {
+    bufferlist data;
+    crypt->decrypt(cache, 0, cache.length(), data, part_ofs);
+    off_t send_size = cache.length() - enc_begin_skip;
+    if (ofs + enc_begin_skip + send_size > end + 1) {
+      send_size = end + 1 - ofs - enc_begin_skip;
+    }
+    res = next->handle_data(data, enc_begin_skip, send_size);
+    enc_begin_skip = 0;
+    ofs += send_size;
+  }
+  return res;
+}
+
+RGWPutObj_BlockEncrypt::RGWPutObj_BlockEncrypt(CephContext* cct,
+                                               RGWPutObjDataProcessor* next,
+                                               std::unique_ptr<BlockCrypt> crypt):
+    RGWPutObj_Filter(next),
+    cct(cct),
+    crypt(std::move(crypt)),
+    ofs(0),
+    cache()
+{
+  block_size = this->crypt->get_block_size();
+}
+
+RGWPutObj_BlockEncrypt::~RGWPutObj_BlockEncrypt() {
+}
+
+int RGWPutObj_BlockEncrypt::handle_data(bufferlist& bl, off_t in_ofs, void **phandle, rgw_obj *pobj, bool *again) {
+  int res = 0;
+  ldout(cct, 25) << "Encrypt " << bl.length() << " bytes" << dendl;
+
+  if (*again) {
+    bufferlist no_data;
+    res = next->handle_data(no_data, in_ofs, phandle, pobj, again);
+    //if *again is not set to false, we will have endless loop
+    //drop info on log
+    if (*again) {
+      ldout(cct, 0) << "*again==true" << dendl;
+    }
+    return res;
+  }
+  off_t bl_ofs = 0;
+  if (cache.length() > 0) {
+      //append before operation.
+      off_t size = block_size - cache.length();
+      if (size > bl.length())
+        size = bl.length();
+      if (size > 0) {
+        char *src = bl.get_contiguous(0, size);
+        cache.append(src,size);
+        bl_ofs += size;
+      }
+      if (cache.length() == block_size) {
+        bufferlist data;
+        crypt->encrypt(cache, 0, block_size, data, ofs);
+        res = next->handle_data(data, ofs, phandle, pobj, again);
+        cache.clear();
+        ofs += block_size;
+        if (res != 0)
+          return res;
+      }
+    }
+  if (bl_ofs < bl.length()) {
+    off_t aligned_size = (bl.length() - bl_ofs) & ~(block_size - 1);
+    //save remainder
+    off_t remainder = (bl.length() - bl_ofs) - aligned_size;
+    if(remainder > 0) {
+      cache.append(bl.get_contiguous(bl_ofs + aligned_size, remainder), remainder);
```

Comment:
is there a requirement for _contiguous_ buffers here? we don't want to force it to reallocate the buffer just to make it contiguous. and passing the resulting `const char*` to `bufferlist::append()` means we have to actually _copy_ the data, when we could otherwise share references to the underlying raw buffers. `bufferlist::copy()` should do the right thing when there are multiple buffers:
```c++
bl.copy(bl_ofs + aligned_size, remainder, cache);
```